src/charm.py

```diff
@@ -1,117 +1,282 @@
+
+        self._tls = TLSRelationService(self.model)
+
         self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.start, self._on_start)
 
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
-        """Define and start a workload using the Pebble API.
+        self.certificates = TLSCertificatesRequiresV3(self, TLS_CERT)
+        self.framework.observe(
+            self.on.certificates_relation_created, self._on_certificates_relation_created
+        )
+        self.framework.observe(
+            self.on.certificates_relation_joined, self._on_certificates_relation_joined
+        )
+        self.framework.observe(
+            self.on.certificates_relation_broken, self._on_certificates_relation_broken
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_available, self._on_certificate_available
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_expiring, self._on_certificate_expiring
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_invalidated, self._on_certificate_invalidated
+        )
+        self.framework.observe(self.on.get_certificate_action, self._on_get_certificate_action)
+        self.framework.observe(
+            self.certificates.on.all_certificates_invalidated,
+            self._on_all_certificates_invalidated,
+        )
 
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
+    def get_hostname(self) -> str:
+        """Get a list containing all ingress hostnames.
 
-        Learn more about interacting with Pebble at at https://juju.is/docs/sdk/pebble.
+        Returns:
+            A list containing service and additional hostnames
+        """
+        # The relation will always exist when this method is called
+        external_hostname = cast(str, get_config(self, "external-hostname"))
+        return external_hostname
 
-        Args:
-            event: event triggering the handler.
+    def _are_relations_ready(self) -> bool:
+        """Check if required relations are ready.
+
+        Returns:
+            Whether required relations are ready and execution should continue.
         """
-        # Get a reference the container attribute on the PebbleReadyEvent
-        container = event.workload
-        # Add initial Pebble config layer using the Pebble API
-        container.add_layer("httpbin", self._pebble_layer, combine=True)
-        # Make Pebble reevaluate its plan, ensuring any services are started if enabled.
-        container.replan()
-        # Learn more about statuses in the SDK docs:
-        # https://juju.is/docs/sdk/constructs#heading--statuses
-        self.unit.status = ops.ActiveStatus()
+        return self._tls.get_tls_relation()
+
+    def _reconcile(self) -> None:
```

Comment:
Just out of curiosity, would this charm be big enough to have a reconcile pattern here?

Reply:
The scope would be close to nginx-ingress-integrator so I'd say i can get quite complex (the scope can grow even bigger too since gateway API offers more features than nginx ingress)